components/sup/src/ctl_gateway/mod.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! CltGateway, short for Control Gateway, is a TCP based client and server connection for
+//! sending and receiving command and control requests to a running Supervisor. Operational tasks
+//! such as starting, stopping, loading, and unloading services are exposed through the
+//! CtlGateway.
+//!
+//! The [`ctl_gateway.client`] and [`ctl_gateway.server`] speak a streaming, multiplexed, binary
+//! protocol defined in [`ctl_gateway.codec`].
+
+pub mod client;
+pub mod codec;
+pub mod server;
+
+use std::fmt;
+use std::io::{self, Read, Write};
+use std::fs::{self, File};
+use std::net::{Ipv4Addr, SocketAddr};
+use std::path::Path;
+
+use common::ui::UIWriter;
+use depot_client::DisplayProgress;
+use futures::prelude::*;
+use hcore;
+use hcore::util::perm;
+use nanoid;
+
+use error::{Result, Error};
+use protocols;
+
+/// Default listening port for the CtlGateway listener.
+pub const DEFAULT_PORT: u16 = 9632;
+/// Name of file containing the CtlGateway secret key.
+pub const CTL_SECRET_FILENAME: &'static str = "CTL_SECRET";
+/// Time to wait in milliseconds for a client connection to timeout.
+pub const REQ_TIMEOUT: u64 = 10_000;
+/// Length of characters in CtlGateway secret key.
+const CTL_SECRET_LEN: usize = 64;
```

Comment:
How'd you arrive at this value? If I understand the [nanoid docs](https://docs.rs/nanoid/0.2.0/nanoid/), each character is a base-64 digit, so 6 bits of entropy, making this a 384-bit pure random key. That seems like a lot, especially considering `nanoid`'s default is 21 characters (126 bits).

I think this is perfectly functional, I just want do make sure I'm understanding what's going on. Since it's not using the `nanoid` default, that implies to me there's a reason. In a perfect world, I'd like more time to review the `nanoid` source, since I'm not familiar with it. Is there a reason we should trust it?

Reply:
I chose nanoid because it's decently known library for generating cryptographically secure pseudo-random character generator with support in a number of languages. It has an extremely small implementation that you can read in one file.

I went with 64 vs their default of 21 because we exchange and validate once per connection. A larger value wasn't really a concern for me and a human would never be reading/writing it.

Another solution is to depend upon the Os' implementation with https://doc.rust-lang.org/rand/rand/struct.OsRng.html. This would be completely fine for us for now, but we may need to drop the dep as we expand the things that the Supervisor can run on.

Reply:
Ah, I'd never heard of nanoid. I did start to read the implementation and I was a bit confused as to what we were getting out of it since we don't need to use a custom alphabet. For something in a security context, when it's down to vetting [something like this](https://docs.rs/crate/nanoid/0.2.0/source/src/generator.rs) (from the nanoid source):
```rust
pub fn universal(random: fn(usize) -> Vec<u8>, alphabet: &[char], size: usize) -> String {
    let mask = (2 << ((alphabet.len() as f64 - 1.0).ln() / 2.0_f64.ln()) as i64) - 1;
    let step: usize = (1.6_f64 * (mask * size) as f64).ceil() as usize;

    let mut id = String::new();

    loop {
        let bytes = random(step);

        for i in 0..step {
            let byte: usize = bytes[i] as usize & mask;

            if alphabet.len() > byte {
                id.push(alphabet[byte]);

                if id.len() == size {
                    return id;
                }
            }
        }
    }
}
```
Versus using `rand` ([in basically the same way nanoid does](https://docs.rs/crate/nanoid/0.2.0/source/src/random.rs)) and `base64`:
```rust
    let mut rng = rand::OsRng::new().unwrap();
    let mut result = vec![0u8; 48];

    rng.fill_bytes(&mut result);
    
    let b64 = base64::encode(&result);
```
I tend to prefer the latter. I don't see anything actually *wrong* with any of the code though, so take it or leave it.